### PR TITLE
Create Ann Index

### PR DESCRIPTION
**Pull request by tom-jerr**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### Issue(s) number: 
DDL for Vector Index #6095
#### Description:
- Tag Ann Index: 
`CREATE TAG ANNINDEX <index_name> ON <tag_name_list>::(<field_name>) [IF NOT EXISTS] ann_index_params}[COMMENT '<comment>'] `

- Edge Ann Index:
`CREATE EDGE ANNINDEX <index_name> ON <tag_name_list>::(<field_name>) [IF NOT EXISTS] ann_index_params}[COMMENT '<comment>'] `

- Ann Index Params
```cpp
{ANNINDEX_TYPE: "IVF", DIM:128, METRIC_TYPE:"L2", NLIST:3, TRAINSIZE:3}
{ANNINDEX_TYPE: "HNSW", DIM:128, METRIC_TYPE:"L2", MAXDEGREE:15, EFCONSTRUCTION:200, MAXELEMENTS:10000}
```
### Ann Index Manager

- In the Storage Daemon, design a `VectorIndexManager` singleton to manage all Ann Indexes within Storaged.  
- Lifecycle of an Ann Index:  
  - **Create**: Create Ann Index via `CreateTagAnnIndex` request.  
    > Unless deleted, it will always be maintained in memory. It needs to be persisted to disk on exit, and reloaded from disk after the system restarts.  
  - **Use**: Use Ann Index to accelerate queries.  
  - **Delete**: Delete Ann Index via `DropTagAnnIndex` request.  
  - **Update**: Update Ann Index via `UpdateTagAnnIndex` request.  

```cpp
class VectorIndexManager final {
 public:
  static VectorIndexManager& getInstance();
  Status init(meta::IndexManager* indexManager, std::string annIndexPath);
  Status start();
  Status stop();
  void waitUntilStop();
  void notifyStop();
  Status createOrUpdateIndex(GraphSpaceID spaceId,
                             PartitionID partitionId,
                             IndexID indexId,
                             const std::shared_ptr<meta::cpp2::AnnIndexItem>& indexItem);
  StatusOr<std::shared_ptr<AnnIndex>> getIndex(GraphSpaceID spaceId,
                                               PartitionID partitionId,
                                               IndexID indexId);
  Status removeIndex(GraphSpaceID spaceId, PartitionID partitionId, IndexID indexId);
  Status addVectors(GraphSpaceID spaceId,
                    PartitionID partitionId,
                    IndexID indexId,
                    const VecData& vecData);
  StatusOr<SearchResult> searchVectors(GraphSpaceID spaceId,
                                       PartitionID partitionId,
                                       IndexID indexId,
                                       const SearchParams& searchParams);
  std::vector<std::shared_ptr<AnnIndex>> getIndexesByPartition(GraphSpaceID spaceId,
                                                               PartitionID partitionId);
  bool hasIndex(GraphSpaceID spaceId, PartitionID partitionId, IndexID indexId) const;
  Status rebuildIndex(GraphSpaceID spaceId,
                      PartitionID partitionId,
                      IndexID indexId,
                      const std::shared_ptr<meta::cpp2::AnnIndexItem>& indexItem);
};
```
#### Ann Index Item

- The general Index Item defines the index of multiple fields in a single schema, and its index parameters only include `s2_max_level` and `s2_max_cells`, which cannot meet the requirements of Ann Index for creating indexes across the **same-named attributes of multiple schemas**. Therefore, we design a new Ann Index Item.

- Ann Index Item defines the creation of an index across the same-named attributes of multiple schemas, including information about all schemas that need to be indexed. At the same time, a list<binary> is used to store the parameters for creating the ann index.
   > ```cpp
   > ann index params:
   > [IVF/*ann type*/, 128 /*dim*/, L2/*metric type*/, 3/*nlist*/, 3/*train size*/]
   > [HNSW/*ann type*/, 128 /*dim*/, L2/*metric type*/, 16/*max degree*/, 200/*ef construction*/, 100000/*max elements*/]
```cpp
struct IndexItem {
    1: common.IndexID       index_id,
    2: binary               index_name,
    3: common.SchemaID      schema_id
    4: binary               schema_name,
    5: list<ColumnDef>      fields,
    6: optional binary      comment,
    7: optional IndexParams index_params,
}

struct AnnIndexItem {
    1: common.IndexID           index_id,
    2: binary                   index_name,
    3: binary                   prop_name,
    4: list<common.SchemaID>    schema_ids,
    5: list<binary>             schema_names,
    6: list<ColumnDef>          fields,
    7: optional binary          comment,
    8: optional list<binary>    ann_params,
}
```
#### Ann Index Request & Response
- The previous `CreateTagIndexReq` only supported the creation of indexes for a single Schema, while Ann Index needs to support the creation of indexes for multiple Schemas with **identically named attributes**. Therefore, we designed the new `CreateTagAnnIndexReq`.

```cpp
struct CreateTagAnnIndexReq {
    1: common.GraphSpaceID      space_id,
    2: binary                   index_name,
    3: list<binary>             tag_names,
    4: IndexFieldDef            field,
    5: bool                     if_not_exists,
    6: optional binary          comment,
    7: optional list<binary>    ann_params,
}

struct ListTagAnnIndexesResp {
    1: common.ErrorCode         code,
    2: common.HostAddr          leader,
    3: list<AnnIndexItem>	items,
}
struct ListEdgeAnnIndexesResp {
    1: common.ErrorCode         code,
    2: common.HostAddr          leader,
    3: list<AnnIndexItem>    	items,
}
```
#### Create Ann Index Syntax Design
`CREATE [TAG | EDGE] ANNINDEX <index_name> ON <tag_name_list>::(<field_name>) [IF NOT EXISTS] ann_index_params [COMMENT '<comment>']`
```cpp
ann_index_params:
{ANNINDEX_TYPE: "IVF", DIM:128, METRIC_TYPE:"L2", NLIST:3, TRAINSIZE:3}
{ANNINDEX_TYPE: "HNSW", DIM:128, METRIC_TYPE:"L2", MAXDEGREE:15, EFCONSTRUCTION:200, MAXELEMENTS:10000}
```

#### Create Ann Index Plan

1. Graphd generates the plan Start->CreateTagAnnIndex->SubmitJob.

2. During execution, metad first creates the Ann index entry internally. After successful creation, an AdminJob is submitted.

3. Metad packs the job parameters and sends them to Storaged’s AdminTask. Storaged processes these tasks through its internal AdminTaskManager to complete the job. This involves a distributed timing issue:
    > **Meta Client cache update mechanism:** The Storage node’s IndexManager obtains index information through MetaClient,       but MetaClient’s cache is updated by heartbeat cycles:
    > - **Timing issue:** After Meta service creates an index, the Storage node needs to wait until the next heartbeat cycle to see the new index.
     > - **Synchronization issue:** getTagAnnIndex directly reads from the cache. If the cache is not updated yet, it will return IndexNotFound.
- So we use a retry mechanism in Storaged. If multiple retries fail, we directly request data from the meta server to force cache refresh.

image

##### Storaged BuildTagVectorIndexTask Plan

For each Partition, do the following operations:

1. Scan all vector attribute data in KVStore (RocksDB) that match the property name.

2. Store the mapping between Vertex ID/Edge Type and VectorID into the id-vid column family of KVStore.
    > Vertex ID is of type std::string. Here VectorID is calculated by hash, so two mappings need to be stored: VectorID->VertexID and VertexID->VectorID.

image

## Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A
### TODO 
- [ ] TCK TEST
Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/storage/admin/BuildTagVectorIndexTask.cpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+              return sharedItem;
+            }
+          }
+        } else {
+          LOG(WARNING) << "Failed to fetch indexes from meta service: " << allIndexesRet.status();
+        }
+      }
+    }
+
+    // 其他错误直接返回
```
Comment (czpmango) [inline src/storage/admin/BuildTagVectorIndexTask.cpp:75]: Do not use Chinese annotations

File: src/storage/VectorIndexManager.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  Status addVectors(GraphSpaceID spaceId,
+                    PartitionID partitionId,
+                    IndexID indexId,
+                    const VecData& vecData);
+
+  // Search vectors in an index
+  StatusOr<SearchResult> searchVectors(GraphSpaceID spaceId,
+                                       PartitionID partitionId,
+                                       IndexID indexId,
+                                       const SearchParams& searchParams);
```
Comment (czpmango) [inline src/storage/VectorIndexManager.h:81]: Why are these interfaces designed here? Concurrent access to the index might require additional locks

File: src/parser/parser.yy [hunk 3]
```diff
@@ -2752,10 +2767,73 @@ opt_index_field_list
...
+            }
+        } else {
+            delete $4;
+            delete $12;
+            throw nebula::GraphParser::syntax_error(@4, "Invalid index type for IVF parameters");
+        }
+        delete $4;
+        delete $12;
+    }
+    // TODO(LZY): L2 and inner product maybe need to as KEYWORD
```
Comment (czpmango) [inline src/parser/parser.yy:2800]: Fix this TODO

File: src/parser/parser.yy [hunk 4]
```diff
@@ -588,6 +595,14 @@ unreserved_keyword
     | KW_CLEAR              { $$ = new std::string("clear"); }
     | KW_ANALYZER           { $$ = new std::string("analyzer"); }
     | KW_VECTOR             { $$ = new std::string("vector"); }
+    | KW_ANNINDEXTYPE       { $$ = new std::string("annindex_type"); }
```
Comment (czpmango) [inline src/parser/parser.yy:598]: ann_index_type
max_degree
max_elements
train_size

File: src/parser/scanner.lex [hunk 5]
```diff
@@ -333,6 +342,7 @@ LABEL_FULL_WIDTH            {CN_EN_FULL_WIDTH}{CN_EN_NUM_FULL_WIDTH}*
 ";"                         { return TokenType::SEMICOLON; }
 "@"                         { return TokenType::AT; }
 "?"                         { return TokenType::QM; }
+"&"                         { return TokenType::VECTORAND; }
```
Comment (czpmango) [inline src/parser/scanner.lex:345]: What is this for?

File: src/meta/processors/job/JobExecutor.cpp [hunk 6]
```diff
@@ -84,6 +85,10 @@ std::unique_ptr<JobExecutor> JobExecutorFactory::createJobExecutor(const JobDesc
     case cpp2::JobType::STATS:
       ret.reset(new StatsJobExecutor(jd.getSpace(), jd.getJobId(), store, client, jd.getParas()));
       break;
+    case cpp2::JobType::BUILD_TAG_VECTOR_INDEX:
+      ret.reset(new BuildTagVectorIndexJobExecutor(
+          jd.getSpace(), jd.getJobId(), store, client, jd.getParas()));
+      break;
```
Comment (czpmango) [inline src/meta/processors/job/JobExecutor.cpp:91]: add test cases for both node and edge

File: src/storage/admin/BuildTagVectorIndexTask.cpp (hunk 1 above)
Comment (tom-jerr) [reply src/storage/admin/BuildTagVectorIndexTask.cpp:75]: OK

File: src/storage/VectorIndexManager.h (hunk 2 above)
Comment (tom-jerr) [reply src/storage/VectorIndexManager.h:81]: I have fixed the design.

File: src/parser/scanner.lex (hunk 5 above)
Comment (tom-jerr) [reply src/parser/scanner.lex:345]: `Create TAG ANNINDEX tag5_idx ON tag5&tag6::(vec) {ANNINDEX_TYPE:'IVF', DIM:3, METRIC_TYPE:'L2', NLIST:3, TRAINSIZE:3}` Syntax Design like this.

File: src/meta/processors/job/JobExecutor.cpp (hunk 6 above)
Comment (tom-jerr) [reply src/meta/processors/job/JobExecutor.cpp:91]: OK

**Review by czpmango — APPROVED** [review]
Generally LGTM.
Please add some tck test cases in later PRs.
